File: soundnotification/src/main/java/org/bug/soundnotification/DelayCalculator.java
package org.bug.soundnotification;

/**
 * This is where the magic happens
 */
public class DelayCalculator {
	private int increment;
	private int limit;
	private int totalRepeats;
	private int totalDelays;

	private int next;
	private int previous;

	public void init(int start, int increment, int limit) {
		this.increment = increment;
		this.limit = limit;

		next = start;
		previous = 0;
		totalRepeats = 1;
		totalDelays = next;
	}

	public int nextDelay() {
		int delay = next;

		if (limit == 0 || (limit > 0 && totalDelays <= limit) || (limit < 0 && totalRepeats <= Math.abs(limit))) {
			switch (increment) {
				case -1: // double the interval
					next *= 2;
					break;
				case -2: // Fibonacci progression
					int tmp = next;
					next += previous;
					previous = tmp;
					break;
				default: // plain linear progression
					next += increment;
			}
			if (limit > 0) {
				totalDelays += next;
			}
			if (limit < 0) {
				totalRepeats++;
			}
		} else {
			delay = 0;
		}

		return delay;
	}
}

File: soundnotification/src/main/java/org/bug/soundnotification/DelayCalculator.kt
package org.bug.soundnotification

import android.content.SharedPreferences
import android.util.Log
import java.io.Serializable
import kotlin.math.abs

/**
 * This is where the magic happens
 */
class DelayCalculator(prefs: SharedPreferences) : Serializable {
    private var increment = 0
    private var limit = 0
    private var totalRepeats = 1
    private var totalDelays = 0
    private var next = 0
    private var previous = 0

    init {
        increment = prefs.getString("increment_list", "0")!!.toInt()
        limit = prefs.getString("limit_list", "0")!!.toInt()
        totalDelays = prefs.getString("start_list", "2")!!.toInt()

        next = totalDelays
    }

    fun nextDelay(): Int {
        Log.d(javaClass.name, "Calc: $this")
        var delay = next
        if (limit == 0 ||
            limit > 0 && totalDelays <= limit ||
            limit < 0 && totalRepeats <= abs(limit)
        ) {
            when (increment) {
                DOUBLE -> next *= 2
                FIB -> {
                    val tmp = next
                    next += previous
                    previous = tmp
                }
                else -> next += increment
            }
            totalDelays += next
            totalRepeats++
        } else {
            delay = 0
        }

        Log.d(javaClass.name, "Returning $delay")
        return delay
    }

    override fun toString() =
        "DelayCalculator(increment=$increment, limit=$limit, totalRepeats=$totalRepeats, totalDelays=$totalDelays, next=$next, previous=$previous)"
}